app/models/concerns/dirty_associations.rb

```diff
@@ -14,16 +14,30 @@
 module DirtyAssociations
   extend ActiveSupport::Concern
 
+  included do
+    after_save :reset_dirty_cache_state
+  end
   module ClassMethods
     # usage:
     #   class Model
     #     dirty_has_many_associations :organizations, :locations
+
     def dirty_has_many_associations(*args)
+      @@dirty_associations ||= []
+      @@dirty_associations += args
+      dirty_associations = @@dirty_associations
       extension = Module.new do
         args.each do |association|
           association_ids = association.to_s
           association_ids = association_ids.singularize + '_ids' unless association.to_s.end_with?('_ids')
 
+          define_method "reset_dirty_cache_state" do
+            dirty_associations.each do |assoc|
```

Comment:
since you now moved this into a block where association represent the association we define the cache for, you shouldn't be iterating over all dirty associations here. Just define `reset_#{association}_dirty_cache_state` method that only cleans up one attribute, so IMHO it should look like this

```ruby
define_method "reset_#{association}_dirty_cache_state" do
  instance_variable_set("@#{association_ids}_changed", false)
  instance_variable_set("@#{association_ids}_was", nil)
end
```

then the (new) `reset_dirty_cache_state` method should only iterate over associations stored in @@dirt_associations array and call respective `reset_#{association}_dirty_cache_state` methods

Reply:
I have made the change and the tests pass, but I get the below error in the rails log since Taxonomix is a module. @ares 

```
18:49:35 rails.1   | 2019-07-16T18:49:35 [F|app|f48f8ad3] NoMethodError (undefined method `+' for nil:NilClass):
18:49:35 rails.1   | 2019-07-16T18:49:35 [F|app|f48f8ad3]   
18:49:35 rails.1   | 2019-07-16T18:49:35 [F|app|f48f8ad3] app/models/concerns/dirty_associations.rb:38:in `dirty_has_many_associations'
18:49:35 rails.1   |  | app/models/concerns/taxonomix.rb:21:in `block in <module:Taxonomix>'
18:49:35 rails.1   |  | app/models/auth_sources/auth_source_ldap.rb:30:in `include'
18:49:35 rails.1   |  | app/models/auth_sources/auth_source_ldap.rb:30:in `<class:AuthSourceLdap>'
18:49:35 rails.1   |  | app/models/auth_sources/auth_source_ldap.rb:21:in `<top (required)>'
18:49:35 rails.1   |  | app/controllers/api/v2/auth_source_ldaps_controller.rb:6:in `<class:AuthSourceLdapsController>'
18:49:35 rails.1   |  | app/controllers/api/v2/auth_source_ldaps_controller.rb:3:in `<module:V2>'
18:49:35 rails.1   |  | app/controllers/api/v2/auth_source_ldaps_controller.rb:2:in `<module:Api>'
18:49:35 rails.1   |  | app/controllers/api/v2/auth_source_ldaps_controller.rb:1:in `<top (required)>'
18:49:35 rails.1   |  | lib/foreman/middleware/telemetry.rb:10:in `call'
18:49:35 rails.1   |  | lib/foreman/middleware/catch_json_parse_errors.rb:9:in `call'
18:49:35 rails.1   |  | lib/foreman/middleware/logging_context_session.rb:22:in `call'
18:49:35 rails.1   |  | lib/foreman/middleware/logging_context_request.rb:11:in `call'
18:49:35 rails.1   |  | /home/vagrant/katello/lib/katello/prevent_json_parsing.rb:12:in `call'
```

Reply:
Is this still the issue? the trace points to line 38 but I guess in current version it's line 35 that would be failing. How do you reproduce it?

In modules, we may need `mattr_accessor` instead but that would work with classes again (same issue as with `@@`). Perhaps we could use
```
class << self
  attr_accessor :dirty_associations
end
```

that should work both in class and module and keep separate storage even if module is included in the class.